### PR TITLE
Add bug popover to TestRun executions table

### PR DESCRIPTION
**Pull request by asankov**
Also, move `fetchBugDetails` and `assignPopoverData`
to tcms/static/js/bugs.js so that it can be reused

### CODE REVIEW

File: tcms/static/js/bugs.js [hunk 1]
```diff
@@ -14,12 +14,10 @@ function loadBugs(selector, filter) {
...
             {
                 data: null,
-                render: function (data, type, full, meta) {
-                    return `<a href="#bugs" data-toggle="popover" data-html="true"
+                render: (data, type, full, meta) => `<a href="#bugs" data-toggle="popover" data-html="true"
                         data-content="undefined" data-trigger="focus" data-placement="top">
                         <span class="fa fa-info-circle"></span>
-                        </a>`;
-                }
+                        </a>`
```
Comment (atodorov) [inline tcms/static/js/bugs.js:20]: This is unrelated, why change it ?

**Review by atodorov — CHANGES_REQUESTED** [review]
Upon loading the page I see several different RPC errors, the first one is "Bug matching query does not exist".  Then I get a bunch of errors for NoneType object not having a `.details()` method coming from URLs marked as bugs for which bug trackers aren't defined in the DB.

File: tcms/static/js/bugs.js (hunk 1 above)
Comment (asankov) [reply tcms/static/js/bugs.js:20]: because it's inconsistent with the other parts of the file

**Comment by asankov** [comment]
> Upon loading the page I see several different RPC errors, the first one is "Bug matching query does not exist". 

will try to reproduce that

> Then I get a bunch of errors for NoneType object not having a .details() method coming from URLs marked as bugs for which bug trackers aren't defined in the DB.

this is because you have created a bug, whose `baseURL` does not correspond to an issue tracker. I am not sure if I should do anything to handle that as part of this PR, or maybe I don't understand the use cases. aren't you supposed to always have a valid bug tracker associated with the bug

**Comment by atodorov** [comment]
> this is because you have created a bug, whose baseURL does not correspond to an issue tracker.

Yes, I understand that. This would be a fairly normal use-case.

> I am not sure if I should do anything to handle that as part of this PR

the error isn't directly related to the current PR, it has been probably happening forever but now is easier to reproduce.

> aren't you supposed to always have a valid bug tracker associated with the bug

Not necessarily. A valid BugSystem definition is only needed of you want to use the integration. If you are happy with just having the links around then the application shouldn't trow a bunch of errors in your face. A very real example is a tester hitting a test failure caused by issues/bugs in tooling outside their control, often times known issues for which there is a ticket reported. We've got a bunch of these right now posted on various 3rd party repos over GitHub.   You certainly don't want and sometimes even can't integrate with 3rd parties but you can always put a link to that inside a TestExecution.

**Comment by asankov** [comment]
I put error checks for the non-existing trackers. 

I was able to reproduce "Bug matching query does not exist". It happens when I have Kiwi TCMS issue tracker on `http://localhost:8000` but I also have a link that is say `http://localhost:8000/bugs/1/`. In this case if finds and issue tracker, but can't find bug with pk=1 in it. I also put additional checks in the tracker class. Let me know if you think this is not the way to go here

File: tcms/testruns/templates/testruns/get.html [hunk 2]
```diff
@@ -263,7 +263,7 @@ <h5><span class="test-executions-count"></span> Results</h5>
                                                                 class="bug-tooltip"
                                                                 style="visibility: hidden;"
                                                                 data-toggle="popover" data-html="true"
-                                                                data-content="undefined"
+                                                                data-content=""
```
Comment (atodorov) [inline tcms/testruns/templates/testruns/get.html:266]: Bring this back on. If something fails users will see "undefined" and are more likely to report a bug.

File: tcms/static/js/bugs.js [hunk 3]
```diff
@@ -55,6 +55,9 @@ function fetchBugDetails(source, popover, cache = bugDetailsCache) {
     }
 
     jsonRPC('Bug.details', [source.href], data => {
+        if (!data) {
+            return
+        }
```
Comment (atodorov) [inline tcms/static/js/bugs.js:60]: Does anything break without this if statement when data is None (what is the equivalent of None in JavaScript) ? I think we're fine without this check.

File: tcms/rpc/api/bug.py [hunk 4]
```diff
@@ -37,6 +37,8 @@ def details(url, **kwargs):
     if not result:
         request = kwargs.get(REQUEST_KEY)
         tracker = tracker_from_url(url, request)
+        if not tracker:
+            return None
```
Comment (atodorov) [inline tcms/rpc/api/bug.py]: To answer my question below I think here you should return a dict which has all of the necessary keys but empty values so that clients won't have to do their own checking. If you insist on returning None it should be documented and then tested.

File: tcms/issuetracker/kiwitcms.py [hunk 5]
```diff
@@ -34,7 +34,12 @@ def details(self, url):
             Provide more details from our own bug tracker!
         """
         bug_id = self.bug_id_from_url(url)
-        bug = Bug.objects.get(pk=bug_id)
+        bug = Bug.objects.filter(pk=bug_id)
```
Comment (atodorov) [inline tcms/issuetracker/kiwitcms.py:37]: filter().first() will return object or none.

Comment (atodorov) [reply tcms/issuetracker/kiwitcms.py:37]: The change from this commit, together with the previous one needs tests and can be independently pushed to master.

File: tcms/rpc/api/bug.py [hunk 6]
```diff
@@ -34,14 +34,16 @@ def details(url, **kwargs):
...
+        return result
 
+    request = kwargs.get(REQUEST_KEY)
+    tracker = tracker_from_url(url, request)
+    if not tracker:
+        return None
+
+    result = tracker.details(url)
+    cache.set(url, result)
     return result
```
Comment (atodorov) [inline tcms/rpc/api/bug.py:47]: This is probably  a very minor improvement on the code climate score and can go together with the other 2 commits in master and their respective tests.

**Review by atodorov — CHANGES_REQUESTED** [review]
Requesting to move the last 3 commits to master, with changes + tests needed b/c testing in this branch is currently botched. Open a PR from a branch in the main repo instead of your own fork in order for the integration tests to have access to their secrets.

File: tcms/static/js/bugs.js (hunk 3 above)
Comment (asankov) [reply tcms/static/js/bugs.js:60]: the equivalent of `None` in JS is `null`, `undefined`, `0`, `""`, etc. Every one of these values will return false and trigger the early exit in this case. Here, if the API returns `None` the result will be parsed as `undefined`. To answer your question, I am not sure if anything will break if we omit this check, will try it out

File: tcms/rpc/api/bug.py (hunk 4 above)
Comment (asankov) [reply tcms/rpc/api/bug.py]: I don't insist on returning `None`. How about an empty dict - `{}`?

**Comment by asankov** [comment]
PR to master - #1879

File: tcms/rpc/api/bug.py (hunk 4 above)
Comment (atodorov) [reply tcms/rpc/api/bug.py]: `assignPopoverData` in the JS source will try using the keys from this dict and IDK how it will break if the result is None or empty dict b/c for the user there won't the the keys it expects.

Comment (asankov) [reply tcms/rpc/api/bug.py]: from the browser console:
```
>> const emptyDict = {}
>> emptyDict.something;
undefined
```
```
>> const undef = undefined;
>> undef.something;
Uncaught TypeError: undef is undefined
```
accessing not existing field on an object/dict, will just return undefined, hence returning an empty should not break anything. I think that even if we return `None` it gets serialized as `{}`, so there should be no issues here.

File: tcms/testruns/templates/testruns/get.html [hunk 7]
```diff
@@ -256,7 +256,19 @@ <h5><span class="test-executions-count"></span> Results</h5>
                                                 <div>
                                                     <ul class="test-execution-hyperlinks"></ul>
                                                     <template id="link-entry">
-                                                        <li><span class="link-icon"></span><a class="link-url" href="${link.url}">${link.name || link.url}</a></li>
+                                                        <li>
+                                                            <span class="link-icon"></span>
+                                                            <a class="link-url" href="${link.url}">${link.name || link.url}</a>
```
Comment (atodorov) [inline tcms/testruns/templates/testruns/get.html:261]: `${link.url}` not replaced when rendered. For a bug where I can fetch the details the URL is displayed fine but leads to `http://127.0.0.1:8000/runs/28/$%7Blink.url%7D`

File: tcms/testruns/static/testruns/js/get.js [hunk 8]
```diff
@@ -284,6 +287,9 @@ function renderLink(link) {
     const template = $(linkEntryTemplate.cloneNode(true))
     if (link.is_defect) {
         template.find('.link-icon').addClass('fa fa-bug')
+        const bugTooltip = template.find('.bug-tooltip')
+        bugTooltip.css('visibility', 'visible')
+        fetchBugDetails({ href: link.url }, bugTooltip)
     }
 
     template.find('.link-url').html(link.name || link.url)
```
Comment (atodorov) [inline tcms/testruns/static/testruns/js/get.js:295]: Must also replace the .href attribute, see comment below..

**Review by atodorov — CHANGES_REQUESTED** [review]

File: tcms/testruns/static/testruns/js/get.js [hunk 9]
```diff
@@ -282,11 +285,16 @@ function addLinkToExecution(event, testExecution) {
 function renderLink(link) {
     const linkEntryTemplate = $('#link-entry')[0].content
     const template = $(linkEntryTemplate.cloneNode(true))
+    const linkUrlEl = template.find('.link-url')
     if (link.is_defect) {
-        template.find('.link-icon').addClass('fa fa-bug')
+        linkUrlEl.addClass('fa fa-bug')
```
Comment (atodorov) [inline tcms/testruns/static/testruns/js/get.js:290]: The previous alternative is better IMO b/c it places the bug icon inside the <span> tag preceding the URL, which makes the icon black (default text color) and has some margin between the icon and the text.

With the current implementation the icon is blue and there is absolutely no margin. Doesn't look good IMO but also renders the link-icon span element redundant.

**Review by atodorov — CHANGES_REQUESTED** [review]

File: tcms/testruns/static/testruns/js/get.js [hunk 10]
```diff
@@ -284,9 +287,14 @@ function renderLink(link) {
     const template = $(linkEntryTemplate.cloneNode(true))
     if (link.is_defect) {
         template.find('.link-icon').addClass('fa fa-bug')
+        const bugTooltip = template.find('.bug-tooltip')
+        bugTooltip.css('visibility', 'visible')
+        fetchBugDetails({ href: link.url }, bugTooltip)
```
Comment (atodorov) [inline tcms/testruns/static/testruns/js/get.js:292]: Missing 3rd parameter which is for page level caching, see usage in `static/js/bugs.js`.

Also the way `renderLink()` calls `fetchBugDetails()` causes all details for all bugs to be loaded during the initial page load while on the TestCase page this is done only when the info icon is clicked. Bug trackers could be slow systems so we'd like to avoid loading data which may not even be shown. 

There are also users who have the bad practice of having between 1500-5000 test cases inside their test plans/test runs while a tester rarely will work with more than 10-20 (they usually split this list between people).

See the usage of `fetchBugDetails()` in bugs.js and try binding it to an on-show/on-click event instead.

**Review by atodorov — CHANGES_REQUESTED** [review]

File: tcms/testruns/static/testruns/js/get.js (hunk 10 above)
Comment (asankov) [reply tcms/testruns/static/testruns/js/get.js]: > Missing 3rd parameter which is for page level caching, see usage in static/js/bugs.js.

see https://github.com/kiwitcms/Kiwi/pull/1873/files#diff-4ff0c8d0df45ac317c8f001b07ae4f11R51

I made the third parameter to default to `bugDetailsCache` in `bugs.js`. This way the consumer should not pass it every time and different pages can share the cache.

File: tcms/static/js/bugs.js [hunk 11]
```diff
@@ -49,3 +47,21 @@ function loadBugs(selector, filter) {
...
+function fetchBugDetails(source, popover, cache = bugDetailsCache) {
+    if (source.href in cache) {
+        assignPopoverData(source, popover, cache[source.href]);
+        return;
+    }
+
+    jsonRPC('Bug.details', [source.href], data => {
+        cache[source.href] = data;
+        assignPopoverData(source, popover, data);
+    }, true);
```
Comment (asankov) [inline tcms/static/js/bugs.js:60]: @atodorov is there a valid reason for this to originally be `true`. we do nothing after this call, so it could be async. that way the browser won't complain about blocking tasks in the main thread.

File: tcms/testruns/static/testruns/js/get.js (hunk 10 above)
Comment (atodorov) [reply tcms/testruns/static/testruns/js/get.js]: > This way the consumer should not pass it every time and different pages can share the cache.

I'm not sure if this will work. If you have 2 pages loaded in parallel they will both load the bugs.js file and create a local (to them) bugsCache variable. So doesn't look like there will be any cache sharing. 

Note: I didn't see the default parameter but I guess the remaining invocations of this function should be adjusted to make use of it.

File: tcms/static/js/bugs.js (hunk 11 above)
Comment (atodorov) [reply tcms/static/js/bugs.js:60]: Probably because the original usage was assigned to an on-show event and when this is executed async it finishes after the popover is shown.

File: tcms/testruns/static/testruns/js/get.js (hunk 10 above)
Comment (atodorov) [reply tcms/testruns/static/testruns/js/get.js]: Note: The previous comment about cache doesn't address the fact that bug details are loaded unconditionally for every bug listed on the page regardless of the fact if the user will ever click on the info icon.

Comment (asankov) [reply tcms/testruns/static/testruns/js/get.js]: I know, I adressed that - https://github.com/kiwitcms/Kiwi/pull/1873/files#diff-e35e7234b4311cd0d2d5fa2bfef17fb1R309-R311

**Review by atodorov — APPROVED** [review]
